pallets/liquid-staking-two/src/protocol.rs

```diff
@@ -0,0 +1,140 @@
+    // after update exchangerate (record_reward),
+    // and then mint/deposit xKSM.
+    fn stake(who: &T::AccountId, amount: BalanceOf<T>) -> DispatchResult {
+        //todo reserve, insurance pool
+        T::Currency::transfer(T::StakingCurrency::get(), who, &Self::account_id(), amount)?;
+
+        MatchingQueueByUser::<T>::try_mutate(
+            who,
+            &Self::current_era(),
+            |user_ledger_per_era| -> DispatchResult {
+                user_ledger_per_era.total_stake_amount = user_ledger_per_era
+                    .total_stake_amount
+                    .checked_add(amount)
+                    .ok_or(ArithmeticError::Overflow)?;
+
+                Ok(())
+            },
+        )?;
+
+        MatchingPoolByEra::<T>::try_mutate(
+            &Self::current_era(),
+            |pool_ledger_per_era| -> DispatchResult {
+                pool_ledger_per_era.total_stake_amount = pool_ledger_per_era
+                    .total_stake_amount
+                    .checked_add(amount)
+                    .ok_or(ArithmeticError::Overflow)?;
+
+                Ok(())
+            },
+        )?;
+        Ok(().into())
+    }
+
+    // After confirmed unbond on relaychain,
+    // and then burn/withdraw xKSM.
+    // before update exchangerate (record_reward)
+    fn unstake(who: &T::AccountId, amount: BalanceOf<T>) -> Result<BalanceOf<T>, DispatchError> {
+        // can not burn directly because we have match mechanism
+        T::Currency::transfer(T::LiquidCurrency::get(), who, &Self::account_id(), amount)?;
+
+        let exchange_rate = ExchangeRate::<T>::get();
+        let asset_amount = exchange_rate
+            .checked_mul_int(amount)
+            .ok_or(Error::<T>::InvalidExchangeRate)?;
+
+        MatchingQueueByUser::<T>::try_mutate(
+            who,
+            &Self::current_era(),
+            |user_ledger_per_era| -> DispatchResult {
+                user_ledger_per_era.total_unstake_amount = user_ledger_per_era
+                    .total_unstake_amount
+                    .checked_add(asset_amount)
+                    .ok_or(ArithmeticError::Overflow)?;
+
+                Ok(())
+            },
+        )?;
+
+        MatchingPoolByEra::<T>::try_mutate(
+            &Self::current_era(),
+            |pool_ledger_per_era| -> DispatchResult {
+                pool_ledger_per_era.total_unstake_amount = pool_ledger_per_era
+                    .total_unstake_amount
+                    .checked_add(asset_amount)
+                    .ok_or(ArithmeticError::Overflow)?;
+
+                Ok(())
+            },
+        )?;
+        Ok(asset_amount)
+    }
+
+    fn claim(who: &T::AccountId) -> DispatchResult {
+        let mut withdrawable_stake_amount = 0u128;
+        let mut withdrawable_unstake_amount = 0u128;
+        let mut remove_record_from_user_queue = Vec::<EraIndex>::new();
+        MatchingQueueByUser::<T>::iter_prefix(who).for_each(|(era_index, user_ledger_per_era)| {
```

Comment:
we can use `fold` here, no need to declare https://github.com/parallel-finance/parallel/pull/362/files#diff-b81993106286609911b130de00ce1f6c1f3b5d7e7f6a71afbfd87989dfde742dR94-R96 and pass to accumulate_claim_by_era

Reply:
Sounds great! But I'm not sure the vector will be cloned or just moved during each step.